_posts/2024-03-22-kraft-migration.md

```diff
@@ -0,0 +1,388 @@
+```
+
+As mentioned, the controller election happens when the first broker creates the ephemeral `/controller` node storing the broker ID.
+
+```shell
+get /controller
+{"version":2,"brokerid":0,"timestamp":"1710691902625","kraftControllerEpoch":-1}
+```
+
+Finally, the `/config/topics` znode lists all the topics in the cluster with the corresponding configuration.
+
+```shell
+get /config/topics/my-topic
+{"version":1,"config":{"retention.ms":"100000"}} // all the other configuration parameters are omitted because using default values
+```
+
+ZooKeeper is designed to be replicated across a number of nodes which built an ensemble.
+A leader node is the one where all the write requests got by clients are forwarded from the other nodes and the operations are coordinated by using the ZooKeeper Atomic Broadcast (ZAB) protocol.
+This protocol keeps all nodes in sync and ensures reliability on messages delivery.
+
+But having ZooKeeper means dealing with a totally different system which needs to be deployed, managed and troubleshooted.
+The admin doesn't have to take care of the Kafka cluster only but also the ZooKeeper ensemble running alongside it.
+Because ZooKeeper requires a quorum, not having the majority of nodes to process requests can also affect the Kafka cluster's availability.
+
+ZooKeeper also represents a bottleneck for scalability and puts a limit on the number of topics and the corresponding partitions supported within a Kafka cluster.
+It has a performance impact when, for example, there is a controller failover and the new elected one has to fetch metadata from ZooKeeper, including all the topics information.
+Also, any metadata update needs to be propagated to the other brokers.
+The problem is that the metadata changes propagations, by using RPCs, grow with the number of topics' partitions involved and more partitions means more metadata to propagate, so it takes longer and the system gets slower.
+
+### Welcome to KRaft!
+
+In order to overcome the limitations related to the ZooKeeper usage, the Kafka community came with the idea of using Kafka itself to store metadata and using the event-driven pattern to take them updated across the nodes.
+The work started with [KIP-500](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500%3A+Replace+ZooKeeper+with+a+Self-Managed+Metadata+Quorum) in late 2019 by introducing a self-managed metadata quorum based on a specific Kafka implementation of the [Raft](https://raft.github.io/) consensus protocol.
+That was named Kafka Raft ... KRaft for friends!
+
+KRaft is an event-based implementation of the Raft protocol with a quorum controller maintaining an event log, a single-partition topic named `__cluster_metadata`, to store the metadata.
```

Comment:
```suggestion
KRaft is an event-based implementation of the Raft protocol with a quorum controller maintaining an event log, a single-partition topic named `__cluster_metadata`, to store the metadata.
Unlike the other topics, this is special one, because records are written to disk synchronously, which is required by the Raft algorithm for correctness.
```